### ISSUE TITLE
[2016 ECCV Oral] SSD: Single Shot MultiBox Detector

### ISSUE THREAD

**Issue by Jasonlee1995** [Jasonlee1995/AI_Papers#12]
Two-shot detector : 성능은 좋지만 speed가 너무 느림 (bounding box proposal, feature resampling 등으로 인한)
Single-shot detector : speed가 빠르지만 성능이 별로임

Real-time이 가능한 single-shot detector을 사용하되, 성능을 올리고자 한 논문

성능은 multi-scale feature map을 통해서 얻어낼 수 있었음

## Model

![image](https://user-images.githubusercontent.com/49643709/114114385-b680e780-991b-11eb-9895-db44af6a63fe.png)
![image](https://user-images.githubusercontent.com/49643709/114120092-d1a52480-9926-11eb-8f3f-82f69dace95f.png)

SSD 모델은 VGG-16 (pre-trained on ILSVRC CLS-LOC dataset)을 backbone으로 사용하며 fc6, fc7 layer를 conv layer로 convert (dropout layer 제거), pool 5 layer를 변경함 (2*2 stride 2 -> 3*3 stride 1 with atrous algorithm)

여기에 추가적으로 convolution layer를 붙여 detection을 진행

## Train

Feature map cell에 다양한 default box가 정해지며, 다음과 같은 방식을 통해 default box를 생성

![image](https://user-images.githubusercontent.com/49643709/114156091-2a8faf80-995d-11eb-9a1a-4f39726bbaef.png)
![image](https://user-images.githubusercontent.com/49643709/114158849-20bb7b80-9960-11eb-8ef8-1c739037fae1.png)

다양한 feature map을 사용하기에, 각 feature map의 default box를 위와 같은 방법을 통해 scaling함 (min=0.2, max=0.9)

또한 default box의 다양한 aspect ratio를 이용하기 위해 width, height 공식을 사용하여 사이즈 조절함

그리고 aspect ratio가 1인 경우, scale을 조금 더 높인 aspect ratio 1의 box를 생성

즉 1개의 feature map location에 aspect ratio가 다른 5가지 + aspect ratio 1인데 사이즈 큰 1가지, 총 6개의 default box를 가짐

생성된 default boxes 중 어떤 default box를 선택할지 다음과 같은 방법을 통해 학습

1. Default box는 multi-scale feature map location에 다양한 aspect ratio로 존재
2. Default box 중 ground truth box의 jaccard overlap이 threshold보다 높은 경우 매칭 (threshold : 0.5)
3. Network로 하여금 위 기준을 만족하는 multiple overlapping default boxes를 predict하도록 학습

당연히 많은 default boxes 중 선택되는 box가 적은 imbalance 문제를 겪기에 confidence loss가 큰 순서로 sorting하여, negative:positive의 비율이 3:1이 되도록 sample하여 loss를 구함 (해당 방법을 적용해야 stable, fast optimize함)

![image](https://user-images.githubusercontent.com/49643709/114155117-15fee780-995c-11eb-953d-5fd4ea738b59.png)
![image](https://user-images.githubusercontent.com/49643709/114155386-67a77200-995c-11eb-85a5-e6879b137762.png)
![image](https://user-images.githubusercontent.com/49643709/114155405-6bd38f80-995c-11eb-896b-c20ba7d853f9.png)

Training objective는 localization loss와 confidence loss의 weighted sum임 (논문에서 cross validation을 통해 alpha=1로 설정)

Localization loss는 smooth L1 loss 사용, confidence loss는 softmax loss임

N은 number of matched default box인데, 0일 경우 loss를 0으로 간주

Data augmentation으로 original image size의 0.1~1 patch sample, aspect ratio로 0.5~2로 변환 후 fixed size로 resize, random horizontal flip을 사용함

## Inference

![image](https://user-images.githubusercontent.com/49643709/114167785-5c5b4300-996a-11eb-9265-a1d2ea298f80.png)

1. Confidence가 threshold보다 낮은 bbox 제외 (threshold=0.01)
2. 각 class에 대하여 jaccard overlap이 0.45 이상인 경우 nms 적용
3. 이미지당 top 200 detection을 keep

## Experiment

SGD lr 0.001, momentum 0.9, weight decay 0.0005에 batch size 32로 학습을 진행

![image](https://user-images.githubusercontent.com/49643709/114161989-9ecd5180-9963-11eb-8b67-0c471926123f.png)
![image](https://user-images.githubusercontent.com/49643709/114166120-4e0c2780-9968-11eb-8071-cd602ea6a6b2.png)
![image](https://user-images.githubusercontent.com/49643709/114166228-6d0ab980-9968-11eb-85ea-20a378d5865e.png)
![image](https://user-images.githubusercontent.com/49643709/114166571-dd193f80-9968-11eb-9582-50611e1aa8b1.png)
![image](https://user-images.githubusercontent.com/49643709/114166602-eb675b80-9968-11eb-8e7d-83812473f320.png)
![image](https://user-images.githubusercontent.com/49643709/114166899-44cf8a80-9969-11eb-9624-5ff624186d84.png)
- new data augmentation : zoom out operation (original image size 16배, mean value로 채워놓은 후 scale한 image 붙인 후 crop)
